contracts/actions/mcd/McdRepayComposite.sol

```diff
@@ -0,0 +1,200 @@
+        emit ActionEvent("McdRepayComposite", logData);
+
+        return bytes32(paybackAmount);
+    }
+
+    /// @inheritdoc ActionBase
+    function executeActionDirect(bytes memory _callData)
+        public
+        payable
+        virtual
+        override(ActionBase, DFSSell, GasFeeTaker)
+    {
+        RepayParams memory repayParams = _parseCompositeParams(_callData);
+        (bytes memory logData, ) = _repay(repayParams);
+        logger.logActionDirectEvent("McdRepayComposite", logData);
+    }
+
+    /// @notice Executes repay logic
+    function _repay(RepayParams memory _repayParams) internal returns (bytes memory logData, uint256 paybackAmount) {
+        assert(_repayParams.exchangeData.destAddr == DAI_ADDR);
+        
+        uint256 ratioBefore;
+        // is part of strategy so check before ratio
+        if (_repayParams.gasUsed != 0) {
+            ratioBefore = getRatio(_repayParams.vaultId, 0);
+        }
+
+        uint256 repayAmount = _repayParams.exchangeData.srcAmount;
+
+        // Withdraw collateral if not using flashloan
+        if (_repayParams.flAddress == address(0)) {
+            _withdrawColl(_repayParams, repayAmount);
+        }
+
+        // Sell collateral asset for debt asset
+        (uint256 exchangedAmount, ) = _dfsSell(
+            _repayParams.exchangeData,
+            address(this),
+            address(this),
+            false
+        );
+
+        // Take gas fee if part of strategy
+        if (_repayParams.gasUsed != 0) {
+            paybackAmount = _takeFee(
+                GasFeeTakerParams(_repayParams.gasUsed, DAI_ADDR, exchangedAmount, MAX_DFS_FEE)
+            );
+        } else {
+            paybackAmount = exchangedAmount;
+        }
+
+        // Payback debt
+        {
+            // if paybackAmount is higher than current debt, repay all debt and send remaining dai to proxy owner
+            (address urn, bytes32 ilk) = getUrnAndIlk(MCD_MANAGER_ADDR, _repayParams.vaultId);
+            uint256 debt = getAllDebt(address(vat), urn, urn, ilk);
+            if (paybackAmount > debt) {
+                DAI_ADDR.withdrawTokens(IDSProxy(address(this)).owner(), paybackAmount - debt);
+                paybackAmount = debt;
+            }
+
+            DAI_ADDR.approveToken(DAI_JOIN_ADDR, paybackAmount);
+            IDaiJoin(DAI_JOIN_ADDR).join(urn, paybackAmount);
+            uint256 daiVatBalance = vat.dai(urn);
+            int256 paybackAmountNormalized = normalizePaybackAmount(
+                address(vat),
+                daiVatBalance,
+                urn,
+                ilk
+            );
+
+            IManager(MCD_MANAGER_ADDR).frob(_repayParams.vaultId, 0, paybackAmountNormalized);
+        }
+
+        if (_repayParams.flAddress != address(0)) {
+            _withdrawColl(_repayParams, repayAmount);
+            _repayParams.exchangeData.srcAddr.withdrawTokens(_repayParams.flAddress, repayAmount);
```

Comment:
this would only work with flashloan providers that don't have any fee, how do we handle if we use a provider with fee? Boost is maybe ok because we flashloan DAI and there should always be free DAI fl providers, but we can't say that for certain for all of coll types